src/quote/Quote.ts

```diff
@@ -0,0 +1,125 @@
+ *
+ * @module Quote
+ * @preferred
+ */
+
+/**
+ * Dummy comment needed for correct doc display, do not remove.
+ */
+
+import Ajv from 'ajv'
+import QuoteSchema from './QuoteSchema'
+import Identity from '../identity/Identity'
+import IQuote, { IQuoteAgreement, IQuoteAttesterSigned } from '../types/Quote'
+import { hashObjectAsStr, verify } from '../crypto/Crypto'
+
+export function validateQuoteSchema(
+  schema: object,
+  validate: object,
+  messages?: string[]
+): boolean {
+  const ajv = new Ajv()
+  ajv.addMetaSchema(QuoteSchema)
+  const result = ajv.validate(
+    JSON.parse(JSON.stringify(schema)),
+    JSON.parse(JSON.stringify(validate))
+  )
+  if (!result && ajv.errors) {
+    if (messages) {
+      ajv.errors.forEach((error: any) => {
+        messages.push(error.message)
+      })
+    }
+  }
+  return !!result
+}
+
+export function verifyQuoteHash(quote: IQuote, quoteHash: string): boolean {
+  return hashObjectAsStr(quote) === quoteHash
+}
+
+export function fromAttesterSignedInput(
+  deserializedQuote: IQuoteAttesterSigned
+): IQuoteAttesterSigned {
+  const { attesterSignature, ...quoteWithHash } = deserializedQuote
+  if (
+    !verify(
+      JSON.stringify(quoteWithHash),
+      attesterSignature,
+      deserializedQuote.attesterAddress
+    )
+  ) {
+    throw Error(
+      `attestersSignature ${deserializedQuote.attesterSignature}
+        does not check out with the supplied data`
+    )
+  }
+  const { quoteHash, ...basicQuote } = quoteWithHash
+  if (!validateQuoteSchema(QuoteSchema, basicQuote)) {
+    throw new Error('Quote does not correspond to schema')
+  }
+  if (!verifyQuoteHash(basicQuote, quoteHash)) {
+    throw Error('Invalid Quote Hash')
+  }
+
+  return {
+    ...basicQuote,
+    quoteHash,
+    attesterSignature,
+  }
+}
+
+export function createAttesterSignature(
+  quoteInput: IQuote,
+  attesterIdentity: Identity
+): IQuoteAttesterSigned {
+  const quoteHash = hashObjectAsStr(quoteInput)
+  const quoteWithHash = { ...quoteInput, quoteHash }
+  const signature = attesterIdentity.signStr(JSON.stringify(quoteWithHash))
+  if (!verifyQuoteHash(quoteInput, quoteHash)) {
+    throw Error('Invalid Quote Hash')
+  }
+  return {
+    ...quoteWithHash,
+    attesterSignature: signature,
+  }
+}
+
+export function fromQuoteDataAndIdentity(
+  quoteInput: IQuote,
+  identity: Identity
+): IQuoteAttesterSigned {
+  if (!validateQuoteSchema(QuoteSchema, quoteInput)) {
+    throw new Error('Quote does not correspond to schema')
+  }
+  return createAttesterSignature(quoteInput, identity)
+}
+
+export function createAgreedQuote(
+  claimerIdentity: Identity,
+  attesterSignedQuote: IQuoteAttesterSigned,
+  requestRootHash: string
+): IQuoteAgreement {
+  const { attesterSignature, ...noAttesterSignature } = attesterSignedQuote
+  if (
+    !verify(
+      JSON.stringify({ ...noAttesterSignature }),
```

Comment:
couldn't we just put in `noAttesterSignature` without the unpacking?

Reply:
Yes, we can. I have updated accordingly.